bitcoin/src/psbt/mod.rs

```diff
@@ -72,87 +290,457 @@ impl Psbt {
+                    input.sequence = None;
+                    input.required_time_locktime = None;
+                    input.required_height_locktime = None;
+                }
+
+                for output in psbt_inner.outputs.iter_mut() {
+                    output.script = None;
+                    output.amount = None;
+                }
+
+                Psbt::new(psbt_inner)
+            }
+        }
+    }
+
+    /// Converts this Psbt into a PsbtV2
+    pub fn get_v2(self) -> Result<Self, Error> {
+        match self.inner.version {
+            Version::PsbtV2 => Ok(self),
+            Version::PsbtV0 => {
+                let unsigned_tx = self.inner.unsigned_tx.unwrap();
+                let mut psbt_inner = PsbtInner {
+                    unsigned_tx: None,
+                    version: Version::PsbtV2,
+                    tx_version: Some(unsigned_tx.version),
+                    fallback_locktime: Some(unsigned_tx.lock_time),
+                    // No information about TxModifiable flags is available in PsbtV0
+                    ..self.inner
+                };
+
+                for (input, txin) in psbt_inner.inputs.iter_mut().zip(unsigned_tx.input.into_iter())
+                {
+                    input.previous_tx_id = Some(txin.previous_output.txid);
+                    input.output_index = Some(txin.previous_output.vout);
+                    input.sequence = Some(txin.sequence);
+
+                    // The following information is not available in PsbtV0
+                    input.required_time_locktime = None;
+                    input.required_height_locktime = None;
+                }
+
+                for (output, txout) in
+                    psbt_inner.outputs.iter_mut().zip(unsigned_tx.output.into_iter())
+                {
+                    output.script = Some(txout.script_pubkey);
+                    output.amount = Some(txout.value);
+                }
 
-        for (vin, psbtin) in tx.input.iter_mut().zip(self.inputs.into_iter()) {
+                Psbt::new(psbt_inner)
+            }
+        }
+    }
+
+    /// Extracts the [`Transaction`] from a PSBT by filling in the available signature information.
+    pub fn extract_tx(self) -> Result<Transaction, Error> {
+        let mut tx = match self.inner.version {
+            Version::PsbtV0 => self.inner.unsigned_tx.unwrap(),
+            _ => self.construct_unsigned_tx()?,
+        };
+
+        for (vin, psbtin) in tx.input.iter_mut().zip(self.inner.inputs.into_iter()) {
             vin.script_sig = psbtin.final_script_sig.unwrap_or_default();
             vin.witness = psbtin.final_script_witness.unwrap_or_default();
         }
 
-        tx
+        Ok(tx)
     }
 
     /// Combines this [`Psbt`] with `other` PSBT as described by BIP 174.
     ///
     /// In accordance with BIP 174 this function is commutative i.e., `A.combine(B) == B.combine(A)`
     pub fn combine(&mut self, other: Self) -> Result<(), Error> {
```

Comment:
`Combine` is no longer a V2 Role, making me think we may not want to expose this on the new interface.

Are we certain `Combine(fA(psbt), fB(psbt)) == fA(fB(psbt)) == fB(fA(psbt)) ` as spec dictates?

Reply:
> `Combine` is no longer a V2 Role, making me think we may not want to expose this on the new interface.

I am not sure about this, better to say I am confused over the Combiner. In BIP 370, there is no mention of Combiner in the Roles section, it says, "PSBTv2 introduces new roles and modifies some existing roles". So, does it mean Combiner doesn't exist for V2 anymore? Also, in the "Unique Identification" section, there is a line -
```
Since PSBT_IN_SEQUENCE can be changed by **Updaters** and **Combiners**, the sequence number in this unsigned transaction must be set to 0 (not final, nor the sequence in PSBT_IN_SEQUENCE).
```

Moreover, in the author's [implementation](https://github.com/bitcoin/bitcoin/pull/21283/files#diff-e8e40219a2efc439389b5a86646cc170c5eeea3d36d0d59f6244fbf4dd46c840L24), V2 `Merge` is implemented.

> Are we certain Combine(fA(psbt), fB(psbt)) == fA(fB(psbt)) == fB(fA(psbt))  as spec dictates?

Yeh, for psbtv0 it is tested, but I have not added tests for other cases. Will add them. Thanks.